app/controllers/actions/JourneyAction.scala

```diff
@@ -31,34 +32,56 @@ import uk.gov.hmrc.play.http.HeaderCarrierConverter
+      case Some(id) => verifyDeclaration(id, request, (d: models.ExportsDeclaration) => types.isEmpty || types.contains(d.`type`))
+      case None     => Future.successful(redirectToRoot(s"Could not obtain the declaration's id for eori ${request.user.eori}"))
     }
-  }
 
-  override def refine[A](request: AuthenticatedRequest[A]): Future[Either[Result, JourneyRequest[A]]] =
-    refiner(request, Seq.empty[DeclarationType])
+  private def refineOnAdditionalTypes[A](request: AuthenticatedRequest[A], additionalTypes: Seq[AdditionalDeclarationType]): RefineResult[A] =
+    request.declarationId match {
+      case Some(id) => verifyDeclaration(id, request, additionalTypes.isEmpty || _.additionalDeclarationType.exists(additionalTypes.contains))
+      case None     => Future.successful(redirectToRoot(s"Could not obtain the declaration's id for eori ${request.user.eori}"))
+    }
 
-  def apply(type1: DeclarationType, others: DeclarationType*): ActionRefiner[AuthenticatedRequest, JourneyRequest] = apply(others.toSeq.+:(type1))
+  override def refine[A](request: AuthenticatedRequest[A]): RefineResult[A] =
+    refineOnDeclarationTypes(request, Seq.empty[DeclarationType])
+
+  def apply(type1: DeclarationType, others: DeclarationType*): ActionRefiner[AuthenticatedRequest, JourneyRequest] =
+    apply(others.toSeq.+:(type1))
 
   def apply(types: Seq[DeclarationType]): ActionRefiner[AuthenticatedRequest, JourneyRequest] =
     new ActionRefiner[AuthenticatedRequest, JourneyRequest] {
-      override protected def refine[A](request: AuthenticatedRequest[A]): Future[Either[Result, JourneyRequest[A]]] = refiner(request, types)
       override protected def executionContext: ExecutionContext = exc
+
+      override protected def refine[A](request: AuthenticatedRequest[A]): RefineResult[A] = refineOnDeclarationTypes(request, types)
     }
+
+  def onAdditionalTypes(additionalTypes: Seq[AdditionalDeclarationType]): ActionRefiner[AuthenticatedRequest, JourneyRequest] =
+    new ActionRefiner[AuthenticatedRequest, JourneyRequest] {
+      override protected def executionContext: ExecutionContext = exc
+
+      override protected def refine[A](request: AuthenticatedRequest[A]): RefineResult[A] =
+        refineOnAdditionalTypes(request, additionalTypes)
+    }
+
+  private def redirectToRoot[A](message: String): Either[Result, JourneyRequest[A]] = {
+    logger.warn(message)
+    Left(Results.Redirect(RootController.displayPage))
+  }
+
+  private def verifyDeclaration[A](id: String, request: AuthenticatedRequest[A], onCondition: ExportsDeclaration => Boolean): RefineResult[A] = {
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+    cacheService.get(id).map {
+      case Some(declaration) =>
+        if (onCondition(declaration)) Right(new JourneyRequest(request, declaration)) else Left(Results.Redirect(RootController.displayPage))
```

Comment:
If `onCondition` is not met (so a Left is returned) the WARN log message `"Could not retrieve from cache, for eori ${request.user.eori}, the declaration with id $id"` is not created (as it used to be in this scenario) :thinking:

Reply:
In truth, the previous log was partially wrong because it was covering 2 different error situations with the same message:
1. If the cache cannot retrieve the declaration by using the id, the error message was correct.
2. If the cache actually retrieve the declaration by using the id but the declaration's type is not the expected one, the message should be different imo.

I'll add a message for the 2nd one btw :+1: